### PR TITLE
initial implementation of single lock addressing #456

### PR DESCRIPTION
**Pull request by buckysballs**
Still needs proper tests but wanted to share
#456

### CODE REVIEW

File: src/main/scala/org/constellation/storage/TransactionService.scala [hunk 1]
```diff
@@ -65,7 +69,13 @@ class TransactionService[F[_]: Sync](dao: DAO)
...
   }
 
-  def pullForConsensus(minCount: Int): F[List[TransactionCacheData]] = {
+  def pullForConsensusSafe(minCount: Int, semaphore: Semaphore[F], roundId: String = "roundId")
+                          (implicit F: Concurrent[F]): F[List[TransactionCacheData]] = {
+    val lock = new SingleLock[F, List[TransactionCacheData]](roundId, semaphore, pullForConsensus(minCount, roundId))
+    lock.acquire
+  }
+
+  def pullForConsensus(minCount: Int, roundId: String = "roundId"): F[List[TransactionCacheData]] = {
```
Comment (Nikolaj-K) [inline src/main/scala/org/constellation/storage/TransactionService.scala]: I see the as of yet unused `roundId` string here. Will this used for some filtering, debug or why is it here too?

File: src/main/scala/org/constellation/consensus/RoundManager.scala [hunk 2]
```diff
@@ -23,10 +24,13 @@ import scala.util.{Failure, Success, Try}
 class RoundManager(config: Config)(implicit dao: DAO) extends Actor with ActorLogging {
   import RoundManager._
 
+  // Needed for getting a Concurrent[IO] instance
+  implicit val ctx = IO.contextShift(ExecutionContext.global)
```
Comment (Nikolaj-K) [inline src/main/scala/org/constellation/consensus/RoundManager.scala]: Scala noob question: Does this variable name spread to everywhere you go with those classes? Generally a 3 letter word with `tx` in it is maybe better to flesh out to a fuller name. If I understand it correctly, the name will not explicitly show up much anyhow.

Comment (marcinwadon) [reply src/main/scala/org/constellation/consensus/RoundManager.scala]: The type is `ContextShift[IO]` that's why `context` (`ctx`). The name isn't so important as far as it is an implicit.

Comment (Nikolaj-K) [reply src/main/scala/org/constellation/consensus/RoundManager.scala]: I haven't fully grasped the scope at which the variable is accessible. Wherever it's accessible, there's the danger that someone might use that name for some sort of transaction, that's why I think a more explicit conflict free name solves it.

File: src/main/scala/org/constellation/storage/TransactionService.scala (hunk 1 above)
Comment (buckysballs) [reply src/main/scala/org/constellation/storage/TransactionService.scala]: Just debugging for now. Still testing

File: src/main/scala/org/constellation/consensus/RoundManager.scala (hunk 2 above)
Comment (buckysballs) [reply src/main/scala/org/constellation/consensus/RoundManager.scala]: @Nikolaj-K you're not wrong about naming for sure, but these are implicits, which are values passed automatically, so it's knowing that the type we need has been invoked is what you'd look for.

Comment (Nikolaj-K) [reply src/main/scala/org/constellation/consensus/RoundManager.scala]: @buckysballs I think I get the use of `ctx`, although the name is still confusingly close to how some particular transaction variable in these classes might end up and confused with. But anyhow, it was just a random suggestions, mostly to show I still read along and try to grasp the code :grin:

File: src/main/scala/org/constellation/primitives/concurrency/SingleLock.scala [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+class SingleLock[F[_], R](name: String, s: Semaphore[F], op: => F[R])(implicit F: Concurrent[F]){
+  def acquire: F[R] =
+    for {
+      x <- s.available
+      _ <- F.delay(println(s"$name >> Availability: $x"))
+      _ <- s.acquire
+      y <- s.available
+      _ <- F.delay(println(s"$name >> Started | Availability: $y"))
+      res <- op
```
Comment (tylerprete) [inline src/main/scala/org/constellation/primitives/concurrency/SingleLock.scala]: I'm not 100% sure -- so I'd recommend testing this -- but I believe that if op returns an error, most implementations of F (cats IO, for example) are going to short-circuit the rest of these commands. Meaning that your semaphore won't get released, and all hell will break loose. 

Again, let's test this to be sure, but I think you'd want to use something like bracket (https://typelevel.org/cats-effect/typeclasses/bracket.html) here.

Comment (buckysballs) [reply src/main/scala/org/constellation/primitives/concurrency/SingleLock.scala]: Oh cool thanks for the tip!

**Pull request merged**